pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,424 @@
+		return reconcile.Result{Requeue: true}, err
+	}
+	if isUnmanaged {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.reconcileS3Resources()
+	if err != nil {
+		r.logger.WithError(err).Error("failed reconciling S3 resources")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *s3EndpointReconciler) reconcileClusterAuthentication() (bool, error) {
+	r.logger.Debugf("reconciling cluster Authentication ServiceAccountIssuer")
+
+	auth, err := r.configclientset.ConfigV1().Authentications().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	serviceAccountIssuer := r.getIssuerURL()
+	if auth.Spec.ServiceAccountIssuer != "" {
+		r.logger.Debug("Authentication ServiceAccountIssuer already specified")
+		if auth.Spec.ServiceAccountIssuer != serviceAccountIssuer {
+			r.logger.Info("Authentication ServiceAccountIssuer is not set to the s3 bucket location, skipping OIDC reconciliation")
+			return true, nil
+		}
+		return false, nil
+	}
+
+	auth.Spec.ServiceAccountIssuer = serviceAccountIssuer
+	_, err = r.configclientset.ConfigV1().Authentications().Update(context.TODO(), auth, metav1.UpdateOptions{})
+	return false, err
+}
+
+func (r *s3EndpointReconciler) getIssuerURL() string {
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s", r.region, r.getBucketName())
+}
+
+func (r *s3EndpointReconciler) getBucketName() string {
+	return fmt.Sprintf("%s-oidc", r.infrastructureName)
+}
+
+func (r *s3EndpointReconciler) reconcileS3Resources() error {
+	r.logger.Debugf("reconciling S3 resources")
+
+	// Get the root secret and create an AWS client
+	secret, err := r.kubeclientset.CoreV1().Secrets(credentialsSecretNamespace).Get(context.TODO(), credentialsSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	accessKey, ok := secret.Data[awsannotator.AwsAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing %s from cloud cred secret", awsannotator.AwsAccessKeyName)
+	}
+
+	secretKey, ok := secret.Data[awsannotator.AwsSecretAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing %s from cloud cred secret", awsannotator.AwsSecretAccessKeyName)
+	}
+
+	awsClient, err := aws.NewClient(accessKey, secretKey, r.region, r.infrastructureName)
+	if err != nil {
+		return fmt.Errorf("error creating aws client: %v", err)
+	}
+
+	// Ensure bucket exists
+	bucketName := r.getBucketName()
+	_, err = awsClient.CreateBucket(&s3.CreateBucketInput{
+		Bucket: awssdk.String(bucketName),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				r.logger.WithField("bucket", bucketName).WithError(err).Error("bucket already exists but it not owned by us")
+				return aerr
+			case s3.ErrCodeBucketAlreadyOwnedByYou:
+				r.logger.WithField("bucket", bucketName).Info("bucket already exists and is owned by us")
```

Comment:
Don't we think this is going to be just a lot of positive logging?

Reply:
yes, moving to Debug